File: Go/Go_100-999/622.go
// Go
// leetcode 622
// https://leetcode.cn/problems/design-circular-queue/

type MyCircularQueue struct {
	Start, Next, ArrSize int
	Arr []int
}


func Constructor(k int) MyCircularQueue {
	return MyCircularQueue{0, 0, k + 1, make([]int, k + 1)}
}


func (t *MyCircularQueue) EnQueue(value int) bool {
	if t.IsFull() {
		return false
	}

	t.Arr[t.Next] = value
	t.Next = (t.Next + 1) % t.ArrSize
	return true
}


func (t *MyCircularQueue) DeQueue() bool {
	if t.IsEmpty() {
		return false
	}

	t.Start = (t.Start + 1) % t.ArrSize
	return true
}


func (t *MyCircularQueue) Front() int {
	if t.IsEmpty() {
		return -1
	}

	return t.Arr[t.Start]
}


func (t *MyCircularQueue) Rear() int {
	if t.IsEmpty() {
		return -1
	}

	return t.Arr[(t.Next - 1 + t.ArrSize) % t.ArrSize]
}


func (t *MyCircularQueue) IsEmpty() bool {
	return t.Start == t.Next
}


func (t *MyCircularQueue) IsFull() bool {
	return t.Start == (t.Next + 1) % t.ArrSize
}


/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * obj := Constructor(k);
 * param_1 := obj.EnQueue(value);
 * param_2 := obj.DeQueue();
 * param_3 := obj.Front();
 * param_4 := obj.Rear();
 * param_5 := obj.IsEmpty();
 * param_6 := obj.IsFull();
 */

File: cpp/cpp_100-999/622.cc
// C++
// leetcode 622
// https://leetcode.cn/problems/design-circular-queue/

class MyCircularQueue {
public:
	int start, next, arrSize;
	vector<int> arr;
    MyCircularQueue(int k_) {
    	arrSize = k_ + 1;
    	arr.resize(arrSize);
    	start = 0;
    	next = 0;
    }
    
    bool enQueue(int value) {
    	if (isFull())
    		return false;
    	arr[next++] = value;
    	next %= arrSize;
        return true;
    }
    
    bool deQueue() {
    	if (isEmpty())
    		return false;
    	start = (start + 1) % arrSize;
    	return true;
    }
    
    int Front() {
    	if (isEmpty())
    		return -1;
    	return arr[start];
    }
    
    int Rear() {
    	if (isEmpty())
    		return -1;
    	return arr[(next + arrSize -1) % arrSize];
    }
    
    bool isEmpty() {
    	return start == next;
    }
    
    bool isFull() {
    	return (next + 1) % arrSize == start;
    }
};

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue* obj = new MyCircularQueue(k);
 * bool param_1 = obj->enQueue(value);
 * bool param_2 = obj->deQueue();
 * int param_3 = obj->Front();
 * int param_4 = obj->Rear();
 * bool param_5 = obj->isEmpty();
 * bool param_6 = obj->isFull();
 */